File: frontend/lib/api/workforce.ts
/**
 * Phase 52: Unified Autonomous Knowledge Worker & Multi-Agent Workforce API Client.
 */

import { api } from './client';

export interface AIWorker {
  id?: string;
  worker_code: string;
  organization_id: string;
  name: string;
  description?: string;
  role: string;
  specialization: string;
  status: 'DRAFT' | 'TESTING' | 'APPROVED' | 'SHADOW' | 'ACTIVE' | 'PAUSED' | 'SUSPENDED' | 'DEPRECATED';
  supervision_level: number;
  agent_version: string;
  model_name: string;
  capabilities: string[];
  total_tasks_completed: number;
  grounding_score: number;
  version: number;
}

export interface AIDepartment {
  id?: string;
  department_code: string;
  name: string;
  description?: string;
  monthly_budget_usd: number;
  current_spend_usd: number;
  status: string;
}

export interface AITeam {
  id?: string;
  team_code: string;
  department_code: string;
  name: string;
  purpose?: string;
  workflow_template: string;
  status: string;
  budget_limit_usd: number;
  members: string[];
}

export interface AIWorkTask {
  id?: string;
  task_code: string;
  objective: string;
  description?: string;
  worker_code?: string;
  priority: string;
  status: string;
  supervision_level: number;
  risk_level: string;
  inputs: Record<string, any>;
  expected_output?: string;
  result_summary?: string;
  dependencies: string[];
  cost_usd: number;
  confidence_score: number;
}

export interface AIHandoff {
  id?: string;
  handoff_code: string;
  from_worker_code: string;
  to_worker_code: string;
  task_code: string;
  context_summary: string;
  artifacts: Array<Record<string, any>>;
  expected_next_action: string;
  confidence_score: number;
  status: string;
}

export interface AIConsensusResult {
  consensus_code: string;
  topic: string;
  participating_workers: string[];
  consensus_score: number;
  has_conflicts: boolean;
  synthesized_conclusion: string;
  dissenting_views: Array<Record<string, any>>;
}

export interface WorkforceOverview {
  status: string;
  total_active_workers: number;
  total_departments: number;
  total_tasks_processed: number;
  pending_human_reviews: number;
  economics: {
    total_tasks_completed: number;
    total_ai_cost_usd: number;
    human_hours_saved: number;
    equivalent_human_cost_usd: number;
    net_cost_savings_usd: number;
    roi_multiple: number;
  };
  active_kill_switches: number;
}

export interface WorkforceCopilotResponse {
  query: string;
  answer: string;
  evidence: string[];
  suggested_actions: string[];
  governance_notice: string;
}

export const workforceApi = {
  getOverview: async (): Promise<WorkforceOverview> => {
    return api.get<WorkforceOverview>('/workforce/overview');
  },

  getWorkers: async (specialization?: string, status?: string): Promise<AIWorker[]> => {
    return api.get<AIWorker[]>('/workforce/workers', {
      params: { specialization, status },
    });
  },

  registerWorker: async (data: {
    name: string;
    role: string;
    specialization: string;
    description?: string;
    supervision_level?: number;
    capabilities?: string[];
  }): Promise<AIWorker> => {
    return api.post<AIWorker>('/workforce/workers', data);
  },

  getDepartments: async (): Promise<AIDepartment[]> => {
    return api.get<AIDepartment[]>('/workforce/departments');
  },

  getTeams: async (departmentCode?: string): Promise<AITeam[]> => {
    return api.get<AITeam[]>('/workforce/teams', {
      params: { department_code: departmentCode },
    });
  },

  decomposeObjective: async (objective: string, domain?: string): Promise<AIWorkTask[]> => {
    return api.post<AIWorkTask[]>('/workforce/tasks/decompose', {
      objective,
      domain: domain || 'GROWTH_EXPANSION',
    });
  },

  getTasks: async (workerCode?: string, status?: string): Promise<AIWorkTask[]> => {
    return api.get<AIWorkTask[]>('/workforce/tasks', {
      params: { worker_code: workerCode, status },
    });
  },

  getPendingReviews: async (): Promise<any[]> => {
    return api.get<any[]>('/workforce/reviews/pending');
  },

  resolveReview: async (reviewId: string, approved: boolean, reviewerId: string, rationale: string): Promise<any> => {
    return api.post<any>(`/workforce/reviews/${reviewId}/resolve`, {
      approved,
      reviewer_id: reviewerId,
      rationale,
    });
  },

  createHandoff: async (data: {
    from_worker_code: string;
    to_worker_code: string;
    task_code: string;
    context_summary: string;
    artifacts: any[];
    expected_next_action: string;
  }): Promise<AIHandoff> => {
    return api.post<AIHandoff>('/workforce/handoffs', data);
  },

  evaluateConsensus: async (topic: string, workerEvaluations: any[]): Promise<AIConsensusResult> => {
    return api.post<AIConsensusResult>('/workforce/consensus', {
      topic,
      worker_evaluations: workerEvaluations,
    });
  },

  triggerKillSwitch: async (targetType: string, targetIdentifier: string, reason: string): Promise<any> => {
    return api.post<any>('/workforce/kill-switch', {
      target_type: targetType,
      target_identifier: targetIdentifier,
      reason,
      operator_id: 'security_admin',
    });
  },

  queryCopilot: async (query: string): Promise<WorkforceCopilotResponse> => {
    return api.post<WorkforceCopilotResponse>('/workforce/copilot/query', { query });
  },
};

File: backend/app/schemas/workforce.py
"""
Pydantic API Schemas for Phase 52:
Unified Autonomous Knowledge Worker & Multi-Agent Workforce Platform.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class WorkerRegisterRequest(BaseModel):
    name: str
    role: str
    specialization: str
    description: Optional[str] = None
    supervision_level: int = 2
    capabilities: Optional[List[str]] = None
    model_name: str = "gemini-1.5-pro"


class WorkerResponse(BaseModel):
    id: Optional[str] = None
    worker_code: str
    organization_id: str
    name: str
    description: Optional[str] = None
    role: str
    specialization: str
    status: str
    supervision_level: int
    agent_version: str
    model_name: str
    capabilities: List[str]
    total_tasks_completed: int
    grounding_score: float
    version: int


class DepartmentResponse(BaseModel):
    id: Optional[str] = None
    department_code: str
    name: str
    description: Optional[str] = None
    monthly_budget_usd: float
    current_spend_usd: float
    status: str


class TeamResponse(BaseModel):
    id: Optional[str] = None
    team_code: str
    department_code: str
    name: str
    purpose: Optional[str] = None
    workflow_template: str
    status: str
    budget_limit_usd: float
    members: List[str]


class TaskDecomposeRequest(BaseModel):
    objective: str
    domain: str = "GROWTH_EXPANSION"


class TaskResponse(BaseModel):
    id: Optional[str] = None
    task_code: str
    objective: str
    description: Optional[str] = None
    worker_code: Optional[str] = None
    priority: str
    status: str
    supervision_level: int
    risk_level: str
    inputs: Dict[str, Any]
    expected_output: Optional[str] = None
    result_summary: Optional[str] = None
    dependencies: List[str]
    cost_usd: float
    confidence_score: float


class HandoffCreateRequest(BaseModel):
    from_worker_code: str
    to_worker_code: str
    task_code: str
    context_summary: str
    artifacts: List[Dict[str, Any]]
    expected_next_action: str
    evidence: Optional[List[str]] = None
    assumptions: Optional[List[str]] = None


class HandoffResponse(BaseModel):
    id: Optional[str] = None
    handoff_code: str
    from_worker_code: str
    to_worker_code: str
    task_code: str
    context_summary: str
    artifacts: List[Dict[str, Any]]
    expected_next_action: str
    confidence_score: float
    status: str


class ConsensusRequest(BaseModel):
    topic: str
    worker_evaluations: List[Dict[str, Any]]


class ConsensusResponse(BaseModel):
    consensus_code: str
    topic: str
    participating_workers: List[str]
    consensus_score: float
    has_conflicts: bool
    synthesized_conclusion: str
    dissenting_views: List[Dict[str, Any]]


class ReviewResolveRequest(BaseModel):
    approved: bool
    reviewer_id: str
    rationale: str


class KillSwitchTriggerRequest(BaseModel):
    target_type: str = "WORKER"  # GLOBAL_WORKFORCE, DEPARTMENT, TEAM, WORKER
    target_identifier: str
    reason: str
    operator_id: str = "security_admin"


class WorkforceCopilotQueryRequest(BaseModel):
    query: str


class WorkforceCopilotQueryResponse(BaseModel):
    query: str
    answer: str
    evidence: List[str]
    suggested_actions: List[str]
    governance_notice: str = "AI Workforce operates under strict supervision levels. Sensitive capabilities require explicit human authorization."
